qemu/qemu.go

```diff
@@ -426,42 +527,70 @@ func (netdev NetDevice) QemuParams(config *Config) []string {
 				deviceParams = append(deviceParams, fmt.Sprintf(",addr=%x", addr))
 			}
 		}
-	}
 
-	netdevParams = append(netdevParams, string(netdev.Type))
-	netdevParams = append(netdevParams, fmt.Sprintf(",id=%s", netdev.ID))
-	netdevParams = append(netdevParams, fmt.Sprintf(",ifname=%s", netdev.IFName))
+		if netdev.DisableModern {
+			deviceParams = append(deviceParams, ",disable-modern=true")
+		}
 
-	if netdev.DownScript != "" {
-		netdevParams = append(netdevParams, fmt.Sprintf(",downscript=%s", netdev.DownScript))
 	}
 
-	if netdev.Script != "" {
-		netdevParams = append(netdevParams, fmt.Sprintf(",script=%s", netdev.Script))
+	// Macvtap can only be connected via fds
+	if (netdev.Type == MACVTAP) && (len(netdev.FDs) == 0) {
+		// Does not make sense to have more Queues that CPUs
+		// Specify the maximum number of queues possible for the
+		// current CPU configuration
+		netdev.FDs, _ = createMacvtapFds(netdev.ID, int(config.SMP.CPUs))
```

Comment:
Are these file descriptors are closed anywhere?  In launcher we close them after the QEMU process has been launched or has failed to launch.  I can see that the FDs aren't leaked but I can't see where they are closed.  We can't expect the user to close them as he doesn't know about them.

Also, any FDs that he assigns to netdev.FDs will be overwritten by this line and will be leaked.

Reply:
Yes. I plan to remove this logic from QEMU. I initially put it here to handle the case where the caller did not do setup properly and we have no means to return error from this API call (which is a bit unfortunate).  But now I feel it is not worth doing. Let the qemu launch fail and we can leave it to the user to recover.

Reply:
@markdryan @sameo it may be worthwhile to provide a way to return error from a lot of these APIs. Yes, this will require changes to the consumers of these packages, but will have benefits in the future.